presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -462,116 +366,265 @@ else if (!result.getBlocked().isDone()) {
+        List<SectionExecution> sectionExecutions = this.sectionExecutions.get(stageId);
+        if (sectionExecutions == null || sectionExecutions.isEmpty()) {
+            return Optional.empty();
+        }
+        return Optional.of(getLast(sectionExecutions));
+    }
+
+    private StageId getStageId(PlanFragmentId fragmentId)
+    {
+        return new StageId(session.getQueryId(), fragmentId.getId());
+    }
+
+    private List<SectionExecution> getOrCreateStageExecutions(List<StreamingPlanSection> sections)
+    {
+        ImmutableList.Builder<SectionExecution> result = ImmutableList.builder();
+        for (StreamingPlanSection section : sections) {
+            StageId sectionId = getStageId(section.getPlan().getFragment().getId());
+            List<SectionExecution> executions = sectionExecutions.computeIfAbsent(sectionId, (ignored) -> new CopyOnWriteArrayList<>());
+            if (!queryStateMachine.isDone() && (executions.isEmpty() || getLast(executions).isFailed())) {
+                PlanFragment sectionRootFragment = section.getPlan().getFragment();
+
+                Optional<int[]> bucketToPartition = Optional.empty();
+                OutputBuffers outputBuffers = createDiscardingOutputBuffers();
+                ExchangeLocationsConsumer locationsConsumer = (fragmentId, tasks, noMoreExchangeLocations) -> {};
+
+                if (isRootFragment(sectionRootFragment)) {
+                    bucketToPartition = Optional.of(new int[1]);
+                    outputBuffers = createInitialEmptyOutputBuffers(sectionRootFragment.getPartitioningScheme().getPartitioning().getHandle())
+                            .withBuffer(new OutputBufferId(0), BROADCAST_PARTITION_ID)
+                            .withNoMoreBufferIds();
+                    OutputBufferId rootBufferId = getOnlyElement(outputBuffers.getBuffers().keySet());
+                    locationsConsumer = (fragmentId, tasks, noMoreExchangeLocations) ->
+                            updateQueryOutputLocations(queryStateMachine, rootBufferId, tasks, noMoreExchangeLocations);
+                }
+
+                int attemptId = executions.size();
+                SectionExecution sectionExecution = sectionStageExecutionFactory.createSectionStageExecutions(
+                        session,
+                        section,
+                        locationsConsumer,
+                        bucketToPartition,
+                        outputBuffers,
+                        summarizeTaskInfo,
+                        remoteTaskFactory,
+                        splitSourceFactory,
+                        attemptId);
+
+                addStateChangeListeners(sectionExecution);
+                executions.add(sectionExecution);
+                result.add(sectionExecution);
+            }
+        }
+        List<SectionExecution> sectionExecutions = result.build();
+        if (queryStateMachine.isDone()) {
+            sectionExecutions.forEach(SectionExecution::abort);
```

Comment:
Where is this part of logic in original code?

Reply:
Previously we created all stage exectuions at once at the beginning.  when the querystatemachine is transitioned to a done state (e.g. failed, canceled, finished early), all remaining tasks would get aborted, and that would cover everything.  Now we might be also creating stage executions in another thread, so after we add the new ones we need to check if the query is finished we should abort them (which is a no-op if they were already aborted)